### PR TITLE
 pod launched by unexpected CNI when the health checking of the agent fails and multus.conf is lost

### PR DESCRIPTION
**Pull request by cyclinder**
#### What type of PR is this?

- release/bug 

**What this PR does / why we need it**:

the prestop of the agent's container would clean up the multus cni config(00-multus.conf), If the container restart but not pod restarts, the config file wouldn't be re-generated, which can cause the multus doesn't work.

the pr moves the multus-cni from init-container to container, and choose the it as the first containers.

**Which issue(s) this PR fixes**:

Fixes https://github.com/spidernet-io/spiderpool/issues/3755

**Special notes for your reviewer**:

1. restart the multus-cni container, the 00-multus.conf would be re-generated.

```
root@spider-worker:/# crictl ps
CONTAINER           IMAGE               CREATED              STATE               NAME                     ATTEMPT             POD ID              POD
a4d9d4af2fe69       d958cfbb222dd       About a minute ago   Running             spiderpool-agent         1                   4ee679bf58463       spiderpool-agent-tzv5j
d831dc97340f0       c0e8690ae66a1       2 minutes ago        Running             multus-cni               2                   4ee679bf58463       spiderpool-agent-tzv5j
8d75c8c8afcc8       0cdd32d9ecad4       6 minutes ago        Running             spiderpool-controller    0                   63e231c0e00a4       spiderpool-controller-6c96859774-xfk8k
1cb443c9e2275       ded66453eb630       24 hours ago         Running             calico-node              0                   07b5a657f52d2       calico-node-99ns6
c6b519fbbc373       ce18e076e9d4b       24 hours ago         Running             local-path-provisioner   0                   84e082ef2ce9e       local-path-provisioner-6f8956fb48-rrjxj
de187f755a123       cbb01a7bd410d       24 hours ago         Running             coredns                  0                   5ccf34350d4d2       coredns-76f75df574-vrvvw
c3bd1ec34ff92       cbb01a7bd410d       24 hours ago         Running             coredns                  0                   0d54dd25c240c       coredns-76f75df574-qxjwr
70065ffc12219       fa4dee78049db       24 hours ago         Running             kube-proxy               0                   d2acd3475056d       kube-proxy-tzcfr
root@spider-worker:/# ls -l /etc/cni/net.d/
total 16
-rw-r--r-- 1 root root  461 Jul 25 08:11 00-multus.conf
-rw-r--r-- 1 root root  752 Jul 24 07:58 10-calico.conflist
-rw------- 1 root root 2737 Jul 25 01:42 calico-kubeconfig
drwxr-xr-x 2 root root 4096 Jul 25 08:11 multus.d
root@spider-worker:/# crictl stop d831dc97340f0
d831dc97340f0
root@spider-worker:/# ls -l /etc/cni/net.d/
total 16
-rw-r--r-- 1 root root  461 Jul 25 08:14 00-multus.conf
-rw-r--r-- 1 root root  752 Jul 24 07:58 10-calico.conflist
-rw------- 1 root root 2737 Jul 25 01:42 calico-kubeconfig
drwxr-xr-x 2 root root 4096 Jul 25 08:14 multus.d
```

2. restart the agent's container, the 00-multus.conf has no changes.

### CODE REVIEW

File: charts/spiderpool/templates/daemonset.yaml [hunk 1]
```diff
@@ -104,37 +104,49 @@ spec:
...
-              mountPath: /host/opt/cni/bin
-              mountPropagation: Bidirectional
-            - name: multus-cfg
-              mountPath: /tmp/multus-conf
-            {{- if .Values.multus.multusCNI.extraVolumes }}
-            {{- include "tplvalues.render" ( dict "value" .Values.multus.multusCNI.extraVolumeMounts "context" $ ) | nindent 12 }}
-            {{- end }}
-        {{- end }}
       containers:
+      {{- if .Values.multus.multusCNI.install }}
```
Comment (weizhoublue) [inline charts/spiderpool/templates/daemonset.yaml:108]: remove the preStop in the spiderpool container

**Comment by weizhoublue** [comment]
currently, there is two containers in the pod, all debugging commands in the CI should change to ` kubectl logs spiderpool-agent -c xxx  ` to get the expected log

File: charts/spiderpool/templates/daemonset.yaml [hunk 2]
```diff
@@ -104,37 +104,49 @@ spec:
...
+          - name: cni-bin-path
+            mountPath: /host/opt/cni/bin
+            mountPropagation: Bidirectional
+          - name: multus-cfg
+            mountPath: /tmp/multus-conf
+          {{- if .Values.multus.multusCNI.extraVolumes }}
+          {{- include "tplvalues.render" ( dict "value" .Values.multus.multusCNI.extraVolumeMounts "context" $ ) | nindent 12 }}
+          {{- end }}
+      {{- end }}
       - name: {{ .Values.spiderpoolAgent.name | trunc 63 | trimSuffix "-" }}
```
Comment (weizhoublue) [inline charts/spiderpool/templates/daemonset.yaml:150]: does it make a difference what if shift this container to the first cone

File: charts/spiderpool/templates/daemonset.yaml (hunk 1 above)
Comment (cyclinder) [reply charts/spiderpool/templates/daemonset.yaml:108]: done

File: charts/spiderpool/templates/daemonset.yaml (hunk 2 above)
Comment (cyclinder) [reply charts/spiderpool/templates/daemonset.yaml:108]: Yes, we should make the agent's container as the first container.

**Comment by weizhoublue** [comment]
@ty-dc  pay an attention to https://github.com/spidernet-io/spiderpool/actions/runs/10094313282/job/27912257643?pr=3758

**Review by weizhoublue — APPROVED** [review]

File: test/Makefile [hunk 3]
```diff
@@ -382,6 +383,15 @@ uninstall_spiderpool:
 	@echo -e "\033[35m [helm uninstall spiderpool] \033[0m"
 	helm uninstall $(RELEASE_NAME) --wait --debug -n $(RELEASE_NAMESPACE) \
 	     --kubeconfig $(E2E_KUBECONFIG)  || { KIND_CLUSTER_NAME=$(E2E_CLUSTER_NAME) ./scripts/debugEnv.sh $(E2E_KUBECONFIG) "detail"   ; exit 1 ; } ; \
+	NODE_LIST=` kind get nodes --name $(E2E_CLUSTER_NAME) `; \
```
Comment (weizhoublue) [inline test/Makefile:386]: I add it in  https://github.com/spidernet-io/spiderpool/pull/3716/files#diff-740cb5a1689091cb894445e46683c255e39689ccba1c6c63d8a8841c8df8817dR411

Comment (cyclinder) [reply test/Makefile:386]: ok,  I will rebase it after https://github.com/spidernet-io/spiderpool/pull/3716 merged.

Comment (weizhoublue) [reply test/Makefile:403]: I mean these codes could be removed

Comment (cyclinder) [reply test/Makefile]: done

**Review by weizhoublue — APPROVED** [review]

**Pull request merged**